### PR TITLE
SDL: Fix shaking with SDL1 and negative offsets

### PR DESCRIPTION
**Pull request by phcoder**
Current code results in a crash

### CODE REVIEW

**Comment by criezy** [comment]
Note that the shake function is documented as taking only positive offsets:
https://github.com/scummvm/scummvm/blob/master/common/system.h#L1155-L1172

So passing negative offsets would be an error in the engine.

That being said, we should indeed avoid a crash. I am unsure about supporting this in some backends though. This may give the impression that negative offsets are supported, but the engine will then break on some other platform. Maybe it would be better to add a sanity check in the implementation of setShakePos to ignore negative offsets and print a warning?

**Comment by phcoder** [comment]
> Note that the shake function is documented as taking only positive offsets: https://github.com/scummvm/scummvm/blob/master/common/system.h#L1155-L1172
> 
> So passing negative offsets would be an error in the engine.
> 
> That being said, we should indeed avoid a crash. I am unsure about supporting this in some backends though. This may give the impression that negative offsets are supported, but the engine will then break on some other platform. Maybe it would be better to add a sanity check in the implementation of setShakePos to ignore negative offsets and print a warning?

This happens in the scumm engine on the day of tentacle on the scene with loud music which is unavoidable for walkthrough. Reading the code I see that scumm uses negative offsets: see method ScummEngine::updateScreenShakeEffect. On the other hand only 2 engines use negative shakes: SCUMM and KYRA

**Comment by criezy** [comment]
> This happens in the scumm engine on the day of tentacle on the scene with loud music which is unavoidable for walkthrough. Reading the code I see that scumm uses negative offsets: see method ScummEngine::updateScreenShakeEffect.

This appears to be a fairly recent change from commit 33d1804b4e. However it is not clear from the commit message and comment added in the code if changing from positive to negative offsets was intended. @athrxx can probably comment.

We could also decide to change the documentation to accept negative offsets if that is needed by some engines. But then we need to also modify all porters as the SDL1 backend may not be the only one not supporting it.

**Comment by athrxx** [comment]
That change to negative offsets was actually intended, since that matches the shaking style of the original code (these offsets that get written into the crt controller registers will move the screen up, not down).

I only ever tested SDL2. I wasn't even aware that SDL1 was still much of a thing for us. With this fix we should probably update the documentation for that method to accept negative offsets.

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp [hunk 1]
```diff
@@ -1181,10 +1189,19 @@ void SurfaceSdlGraphicsManager::internUpdateScreen() {
...
-		SDL_Rect blackrect = {0, 0, (Uint16)(_videoMode.screenWidth * _videoMode.scaleFactor), (Uint16)(_gameScreenShakeYOffset * _videoMode.scaleFactor)};
+		SDL_Rect blackrect = {0, 0, (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
 
-		if (_videoMode.aspectRatioCorrection && !_overlayInGUI)
-			blackrect.h = real2Aspect(blackrect.h - 1) + 1;
+		if (_gameScreenShakeYOffset >= 0)
+			blackrect.h = (Uint16)(_gameScreenShakeYOffset * _videoMode.scaleFactor);
+		else {
+			blackrect.h -= ((-_gameScreenShakeYOffset) * _videoMode.scaleFactor);
+			blackrect.y = ((-_gameScreenShakeYOffset) * _videoMode.scaleFactor);
```
Comment (criezy) [inline backends/graphics/surfacesdl/surfacesdl-graphics.cpp:1198]: And here again I have some doubts on the logic for the black rect size and position in the case of a negative offset. To me you seem to compute the non-black rect instead.

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp [hunk 2]
```diff
@@ -1168,10 +1168,18 @@ void SurfaceSdlGraphicsManager::internUpdateScreen() {
 	// so this isn't needed there.
 	if (_currentShakeXOffset != _gameScreenShakeXOffset ||
 		(_cursorNeedsRedraw && _mouseLastRect.x <= _currentShakeXOffset)) {
-		SDL_Rect blackrect = {0, 0, (Uint16)(_gameScreenShakeXOffset * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
+		SDL_Rect blackrect = {0, 0, (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
```
[the remark this answers is not in the document]
Comment (phcoder) [reply backends/graphics/surfacesdl/surfacesdl-graphics.cpp:1171]: Done

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp [hunk 3]
```diff
@@ -1181,10 +1189,19 @@ void SurfaceSdlGraphicsManager::internUpdateScreen() {
 	}
 	if (_currentShakeYOffset != _gameScreenShakeYOffset ||
 		(_cursorNeedsRedraw && _mouseLastRect.y <= _currentShakeYOffset)) {
-		SDL_Rect blackrect = {0, 0, (Uint16)(_videoMode.screenWidth * _videoMode.scaleFactor), (Uint16)(_gameScreenShakeYOffset * _videoMode.scaleFactor)};
+		SDL_Rect blackrect = {0, 0, (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
```
[the remark this answers is not in the document]
Comment (phcoder) [reply backends/graphics/surfacesdl/surfacesdl-graphics.cpp:1192]: Done

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp [hunk 4]
```diff
@@ -1168,10 +1168,18 @@ void SurfaceSdlGraphicsManager::internUpdateScreen() {
...
 		(_cursorNeedsRedraw && _mouseLastRect.x <= _currentShakeXOffset)) {
-		SDL_Rect blackrect = {0, 0, (Uint16)(_gameScreenShakeXOffset * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
+		SDL_Rect blackrect = {0, 0, (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
 
-		if (_videoMode.aspectRatioCorrection && !_overlayInGUI)
+		if (_gameScreenShakeXOffset >= 0)
+			blackrect.w = (Uint16)(_gameScreenShakeXOffset * _videoMode.scaleFactor);
+		else {
+			blackrect.w -= ((-_gameScreenShakeXOffset) * _videoMode.scaleFactor);
+			blackrect.x = ((-_gameScreenShakeXOffset) * _videoMode.scaleFactor);
```
[the remark this answers is not in the document]
Comment (phcoder) [reply backends/graphics/surfacesdl/surfacesdl-graphics.cpp:1177]: Done

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp (hunk 1 above)
Comment (phcoder) [reply backends/graphics/surfacesdl/surfacesdl-graphics.cpp:1198]: Done

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp [hunk 5]
```diff
@@ -1168,10 +1168,18 @@ void SurfaceSdlGraphicsManager::internUpdateScreen() {
...
-		SDL_Rect blackrect = {0, 0, (Uint16)(_gameScreenShakeXOffset * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
+		SDL_Rect blackrect = {0, 0, (Uint16)(_videoMode.screenWidth * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
 
-		if (_videoMode.aspectRatioCorrection && !_overlayInGUI)
+		if (_gameScreenShakeXOffset >= 0)
+			blackrect.w = (Uint16)(_gameScreenShakeXOffset * _videoMode.scaleFactor);
+		else {
+			blackrect.w = ((-_gameScreenShakeXOffset) * _videoMode.scaleFactor);
+			blackrect.x = ((_videoMode.screenWidth + _gameScreenShakeXOffset) * _videoMode.scaleFactor);
+		}
```
Comment (criezy) [inline backends/graphics/surfacesdl/surfacesdl-graphics.cpp:1178]: A bit of nitpicking, but this whole block could be simplified as follow:
```c++
SDL_Rect blackrect = {0, 0, (Uint16)(ABS(_gameScreenShakeXOffset) * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
if (_gameScreenShakeXOffset < 0)
    blackrect.x = ((_videoMode.screenWidth + _gameScreenShakeXOffset) * _videoMode.scaleFactor);
```

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp [hunk 6]
```diff
@@ -1181,10 +1189,19 @@ void SurfaceSdlGraphicsManager::internUpdateScreen() {
...
+		SDL_Rect blackrect = {0, 0, (Uint16)(_videoMode.screenWidth * _videoMode.scaleFactor), (Uint16)(_videoMode.screenHeight * _videoMode.scaleFactor)};
 
-		if (_videoMode.aspectRatioCorrection && !_overlayInGUI)
-			blackrect.h = real2Aspect(blackrect.h - 1) + 1;
+		if (_gameScreenShakeYOffset >= 0)
+			blackrect.h = (Uint16)(_gameScreenShakeYOffset * _videoMode.scaleFactor);
+		else {
+			blackrect.h = ((-_gameScreenShakeYOffset) * _videoMode.scaleFactor);
+			blackrect.y = ((_videoMode.screenHeight + _gameScreenShakeYOffset) * _videoMode.scaleFactor);
+		}
```
Comment (criezy) [inline backends/graphics/surfacesdl/surfacesdl-graphics.cpp:1199]: A similar simplification to what suggested above for the vertical black rect could also be applied here.

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp [hunk 7]
```diff
@@ -1307,22 +1324,37 @@ void SurfaceSdlGraphicsManager::internUpdateScreen() {
...
+				dst_x += _currentShakeXOffset;
+			else
+				src_x += -_currentShakeXOffset;
+
+			if (_currentShakeYOffset >= 0)
+				dst_y += _currentShakeYOffset;
+			else
+				src_y += -_currentShakeYOffset;
 
 			if (dst_x < width && dst_y < height) {
```
Comment (criezy) [inline backends/graphics/surfacesdl/surfacesdl-graphics.cpp:1346]:
```suggestion
			if (dst_x < width && dst_y < height && src_x < width && src_y < height) {
```

**Review by criezy — COMMENTED** [review]
The code looks mostly good. I have added two minor comments that can be ignored for code simplification, and a sanity check to avoid negative width or height in the case of a negative offset that moves a dirty rect fully outside the screen.

**Review by sev- — APPROVED** [review]
Looks good to me

**Pull request merged**